File: javascript/src/parser.js
import {Element, FountainTitleEntry, 
        FountainAction, FountainHeading, FountainCharacter, 
        FountainDialogue, FountainParenthesis, FountainLyric,
        FountainTransition, FountainPageBreak, FountainNote,
        FountainBoneyard, FountainSection, FountainScript} from "./fountain.js";

// Incremental parser - use .addText(), .addLines(), .addLine() to parse, use .script to retrieve the parsed script.
export class FountainParser {

    constructor() {        
        this.script = new FountainScript();

        this.mergeActions = true;
        this.mergeDialogue = true;

        this._inTitlePage = true;
        this._multiLineHeader = false;
        
        this._lineBeforeBoneyard = "";
        this._boneyard = null;

        this._lineBeforeNote = "";
        this._note = null;

        this._pending = [];
        this._padActions = [];
        
        this._line = "";
        this._lineTrim = "";
        this._lastLineEmpty = true;
        this._lastLine = "";

        this._inDialogue = false;
    }

    // Expects UTF-8 text
    addText(inputText) {

        const lines = inputText.split(/\r?\n|\r/);
        return this.addLines(lines);
    }

    // Expects array of UTF-8 text lines
    addLines(lines) {

        for(const line of lines) {
            this.addLine(line);
        }
        this.finalize();
    }

    // Expects a single UTF-8 text line
    addLine(line) {

        this._lastLine= this._line;
        this._lastLineEmpty = (!this._line.trim());

        this._line = line;

        if (this._parseBoneyard())
            return;

        if (this._parseNotes())
            return;

        this._lineTrim = this._line.trim();

        // Some decisions can't be made until the next line lands
        if (this._pending.length>0)
            this._parsePending();

        if (this._inTitlePage && this._parseTitlePage())
            return;

        if (this._parseSection())
            return;
    
        if (this._parseForcedAction())
            return true;

        if (this._parseForcedSceneHeading())
            return true;

        if (this._parseForcedCharacter())
            return true;

        if (this._parseForcedTransition())
            return true;

        if (this._parsePageBreak())
            return;

        if (this._parseLyrics())
            return;

        if (this._parseCentredText())
            return;

        if (this._parseSceneHeading())
            return;

        if (this._parseTransition())
            return;

        if (this._parseParenthesis())
            return;

        if (this._parseCharacter())
            return;

        if (this._parseDialogue())
            return;

        this._parseAction();
    }

    // If you have definitely finished parsing, call this, as it completes
    // anything that's been waiting for the next line.
    // This is automatically called by addLines() and addText()
    finalize() {
        this._line = "";
        this._parsePending();
    }

    _getLastElem() {
        const elems = this.script.elements;
        if (elems.length>0)
            return elems[elems.length-1];
        return null;
    }
    
    // Adds a new element or merges with existing element
    _addElement(elem) {

        let lastElem = this._getLastElem();

        // Are we trying to add a blank action line?
        if (elem.type == Element.ACTION && elem._text.trim()=="" && !elem.centered) {

            this._inDialogue = false;

            // If this follows an existing action line, put it on as possible padding.
            if (lastElem && lastElem.type == Element.ACTION) {
                this._padActions.push(elem);
                return;
            }
            return;
        }

        // Add padding if there's some outstanding and we're just about to add another action.
        if (elem.type == Element.ACTION && this._padActions.length>0) {

            if (this.mergeActions && !lastElem.centered) {
                for(const padAction of this._padActions) {
                    lastElem._text+="\n"+padAction._text;
                }
            }
            else {
                for(const padAction of this._padActions) {
                    this.script.elements.push(padAction);
                }
            }
        }

        this._padActions = [];

        // If we're allowing actions to be merged, do it here.
        if (this.mergeActions && elem.type == Element.ACTION && !elem.centered) {    
            if (lastElem && lastElem.type == Element.ACTION && !lastElem.centered) {
                lastElem._text+= "\n"+elem._text;
                return;
            }
        }

        this.script.elements.push(elem);

        this._inDialogue = (elem.type == Element.CHARACTER || elem.type == Element.PARENTHESIS || elem.type == Element.DIALOGUE);
    }

    _parsePending() {

        for (const pending of this._pending) {

            if (pending.type == Element.TRANSITION) {

                if (!this._line.trim()) {  // Blank line, so it's definitely a transition
                    this._addElement(pending.element);
                } else {
                    this._addElement(pending.backup);
                }
            } else if (pending.type == Element.CHARACTER) {
                if (this._line.trim()) {  // Filled line, so it's definitely a piece of dialogue
                    this._addElement(pending.element);
                } else {
                    this._addElement(pending.backup);
                }
            }
        }
        this._pending = [];

    }

    _parseTitlePage() {

        const regexTitleEntry = /^\s*([A-Za-z0-9 ]+?)\s*:\s*(.*?)\s*$/;
        const regexTitleMultilineEntry = /^( {3,}|\t)/;

        let match = this._line.match(regexTitleEntry);
        if (match) {    // It's of form key:text
            let text = match[2];
            this.script.headers.push(new FountainTitleEntry( match[1], text) ); 
            this._multiLineHeader = (text.length==0);
            return true

        } 
        
        if (this._multiLineHeader) { // If we're expecting text on this line
            if (regexTitleMultilineEntry.test(this._line)) {
                let header = this.script.headers[this.script.headers.length-1];
                header._text+="\n"+this._line;
                return true;
            }

        }

        this._inTitlePage = false;
        return false;
    }

    _parsePageBreak() {

        const regexPageBreak = /^\s*={3,}\s*$/;
        if (regexPageBreak.test(this._line)) {
            this._addElement(new FountainPageBreak());
            return true;
        }
        return false;
    }

    _parseLyrics() {

        if (this._lineTrim.startsWith('~')) {
            this._addElement(new FountainLyric(this._lineTrim.slice(1).trimStart()));
            return true;
        }
        return false;
    }

    _parseCentredText() {

        if (this._lineTrim.startsWith('>') && this._lineTrim.endsWith('<')) {
            let newElem = new FountainAction(this._lineTrim.slice(1,this._lineTrim.length-1));
            newElem.centered = true;
            this._addElement(newElem);
            return true;
        }
        return false;
    }

    _decodeHeading(line) {
        const regex = /^(.*?)(?:\s*#([a-zA-Z0-9\-.]+)#)?$/;
        const match = line.match(regex);
        return match ? { text: match[1].trim(), sceneNum: match[2] || null } : null;
    }

    _parseForcedSceneHeading() {
        const regex = /^\.[a-zA-Z0-9]/;
        if (regex.test(this._lineTrim)) {
            let heading = this._decodeHeading(this._lineTrim.slice(1));
            this._addElement(new FountainHeading(heading.text, heading.sceneNum, true));
            return true;
        }
        return false;
    }

    _parseSceneHeading() {

        const regexHeading = /^\s*((INT|EXT|EST|INT\.\/EXT|INT\/EXT|I\/E)(\.|\s))|(FADE IN:\s*)/i;
        if (regexHeading.test(this._line)) {
            let heading = this._decodeHeading(this._lineTrim);
            this._addElement(new FountainHeading(heading.text, heading.sceneNum));
            return true;
        }
        return false;
    }

    _parseForcedTransition() {
        if (this._lineTrim.startsWith(">") && !this._lineTrim.endsWith("<")) {
            this._addElement(new FountainTransition(this._lineTrim.slice(1).trim(), true));
            return true;
        }
        return false;
    }

    _parseTransition() {
        const regexTransition = /^\s*(?:[A-Z\s]+TO:)\s*$/;
        if (regexTransition.test(this._line) && !this._lastLine.trim()) {

            if (this._lastLineEmpty) {
                // Can't commit to which this is until we've checked the next line is empty.
                this._pending.push( {
                    type: Element.TRANSITION, 
                    element: new FountainTransition(this._lineTrim),
                    backup: new FountainAction(this._lineTrim)
                } );

                return true;
            }
        }  
        return false; 
    }

    _parseParenthesis() {
       
        const regexParenthesis = /^\s*\((.*)\)\s*$/
        let lastElem = this._getLastElem();
        let match = this._line.match(regexParenthesis);
        if (match && this._inDialogue && lastElem && (lastElem.type==Element.CHARACTER || lastElem.type==Element.DIALOGUE) ) {
            this._addElement(new FountainParenthesis(match[1]));
            return true;
        }
        return false;
    }

    _decodeCharacter(line) {
        // Remove any CONT'D notes
        const regexCont = /\(\s*CONT[’']D\s*\)/g;
        const regexCharacter = /^([^(\^]+?)\s*(?:\((.*)\))?(?:\s*\^\s*)?$/;
        
        let lineTrim = line.trim().replace(regexCont, ""); 
        
        const match = lineTrim.match(regexCharacter);
        if (match) {
            const name = match[1].trim(); // Extract NAME
            const extension = match[2] ? match[2].trim() : null; // Extract extension if present
            const hasCaret = line.trim().endsWith('^'); // Check for the caret
            return { name:name, dual:hasCaret, extension:extension };
        }
        return null; // Invalid format
    }

    _parseForcedCharacter() {
        // Remove any CONT'D notes
        if (this._lineTrim.startsWith("@")) {

            let lineTrim = this._lineTrim.slice(1);

            let character = this._decodeCharacter(lineTrim);
            if (character==null)
                return false;

            this._addElement(new FountainCharacter(lineTrim, character.name, character.extension, character.dual));

            return true;
        }
        return false;
    }

    _parseCharacter() {

        // Remove any CONT'D notes
        const regexCont = /\(\s*CONT[’']D\s*\)/g;
        let lineTrim = this._lineTrim.replace(regexCont, "").trim(); 

        const regexCharacter = /^([A-Z][^a-z]*?)\s*(?:\(.*\))?(?:\s*\^\s*)?$/;
        if (this._lastLineEmpty && regexCharacter.test(lineTrim)) {

            let character = this._decodeCharacter(lineTrim);
            if (character==null)
                return false;

            let charElem = new FountainCharacter(lineTrim, character.name, character.extension, character.dual);

            // Can't commit to which this is until we've checked the next line isn't empty.
            this._pending.push( {
                type: Element.CHARACTER, 
                element: charElem,
                backup: new FountainAction(this._lineTrim)
            } );

            return true;

        }
        return false;
    }

    _parseDialogue() {

        let lastElem = this._getLastElem();
        if (lastElem && this._line.length>0 && (lastElem.type==Element.CHARACTER || lastElem.type==Element.PARENTHESIS)) {
            this._addElement(new FountainDialogue(this._lineTrim));
            return true;
        }

        // Was the previous line dialogue? If so, offer possibility of merge
        if (lastElem && lastElem.type==Element.DIALOGUE) {

            // Special case - line-break in Dialogue. Only valid with more than one white-space character in the line.
            if ( this._lastLineEmpty && this._lastLine.length>0 ) {
                if (this._lastLineEmpty) {
                    if (this.mergeDialogue)
                        lastElem._text+="\n";
                    else
                        this._addElement(new FountainDialogue(""));
                }
                if (this.mergeDialogue)
                    lastElem._text+="\n" + this._lineTrim;
                else
                    this._addElement(new FountainDialogue(this._lineTrim)); 
                return true;
            }

            if (!this._lastLineEmpty && this._lineTrim.length>0) {
                if (this.mergeDialogue)
                    lastElem._text+="\n" + this._lineTrim;
                else
                    this._addElement(new FountainDialogue(this._lineTrim)); 
                return true;
            }
        }

        return false;
    }

    _parseForcedAction() {
        if (this._lineTrim.startsWith("!")) {
            this._addElement(new FountainAction(this._lineTrim.slice(1), true));
            return true;
        }
        return false;
    }

    _parseAction() {
        this._addElement(new FountainAction(this._line));
    }

    // Returns null if there is no content to continue parsing
    _parseBoneyard() {

        // Deal with any in-line boneyards
        let open = this._line.indexOf("/*");
        let close = this._line.indexOf("*/");
        let lastTag = -1;
        while (open>-1 && close>-1) {
            let boneyardText = this._line.slice(open+2, close);
            this.script.boneyards.push(new FountainBoneyard(boneyardText));
            let tag = `/*${this.script.boneyards.length-1}*/`;
            this._line = this._line.slice(0,open)+tag+this._line.slice(close+2);
            lastTag = open+tag.length;
            open = this._line.indexOf("/*", lastTag);
            close = this._line.indexOf("*/", lastTag);
        }

        // If not in boneyard, check for boneyard content
        if (!this._boneyard) {

            let idx = this._line.indexOf("/*", lastTag);
            if (idx>-1) { // Move into boneyard
                this._lineBeforeBoneyard = this._line.slice(0, idx);
                this._boneyard = new FountainBoneyard(this._line.slice(idx+2));
                return true;
            }

        } else {

            // Check for end of note content
            let idx = this._line.indexOf("*/", lastTag);
            if (idx>-1) {
                this._boneyard+="\n"+this._line.slice(0, idx);
                this.script.boneyards.push(this._boneyard);
                let tag = `/*${this.script.boneyards.length-1}*/`;
                this._line = this._lineBeforeBoneyard+tag+this._line.slice(idx+2);
                this._lineBeforeBoneyard = "";
                this._boneyard = null;
            }
            else { // Still in boneyard
                this._boneyard+="\n"+this._line;
                return true;
            }
        }
        return false; 
    }

    // Returns null if there is no content to continue parsing
    _parseNotes() {

        // Deal with any in-line notes
        let open = this._line.indexOf("[[");
        let close = this._line.indexOf("]]");
        let lastTag = -1;
        while (open>-1 && close>-1) {
            let noteText = this._line.slice(open+2, close);
            this.script.notes.push(new FountainNote(noteText));
            let tag = `[[${this.script.notes.length-1}]]`;
            this._line = this._line.slice(0,open)+tag+this._line.slice(close+2);
            lastTag = open+tag.length;
            open = this._line.indexOf("[[", lastTag);
            close = this._line.indexOf("]]", lastTag);
        }

        // If not in notes, check for note content
        if (!this._note) {

            let idx = this._line.indexOf("[[", lastTag);
            if (idx>-1) { // Move into notes
                this._lineBeforeNote = this._line.slice(0, idx);
                this._note = new FountainNote(this._line.slice(idx+2));
                this._line = this._lineBeforeNote;
                return true;
            }

        } else {

            // Check for end of note content
            let idx = this._line.indexOf("]]", lastTag);
            if (idx>-1) {
                this._note+="\n"+this._line.slice(0, idx);
                this.script.notes.push(this._note);
                let tag = `[[${this.script.notes.length-1}]]`;
                this._line = this._lineBeforeNote+tag+this._line.slice(idx+2);
                this._lineBeforeNote = "";
                this._note = null;
            }
            else { // Still in notes
                this._note+="\n"+this._line;
                return true;
            }
        } 
        return false;
    }

    _parseSection() {
        if (this._lineTrim.startsWith("###")) {
            this._addElement(new FountainSection(3, this._lineTrim.slice(3).trim()));
            return true;
        }

        if (this._lineTrim.startsWith("##")) {
            this._addElement(new FountainSection(2, this._lineTrim.slice(2).trim()));
            return true;
        }

        if (this._lineTrim.startsWith("#")) {
            this._addElement(new FountainSection(1, this._lineTrim.slice(1).trim()));
            return true;
        }
        return false;
    }

}

File: cpp/src/parser.cpp
#include "fountain_tools/parser.h"
#include "fountain_tools/utils.h"
#include <sstream>
#include <iostream>

namespace Fountain {

Parser::Parser()
    : script(std::make_shared<Script>()) {}

void Parser::addText(const std::string& inputText) {
    std::istringstream stream(inputText);
    std::string line;
    std::vector<std::string> lines;

    while (std::getline(stream, line)) {
        lines.push_back(line);
    }

    addLines(lines);
}

void Parser::addLines(const std::vector<std::string>& lines) {
    for (const auto& line : lines) {
        addLine(line);
    }
    finalizeParsing();
}

void Parser::addLine(const std::string& inputLine) {
    lastLine = line;
    lastLineEmpty = isWhitespaceOrEmpty(line);

    line = inputLine;

    if (parseBoneyard() || parseNotes()) return;

    lineTrim = trim(line);

    if (!pending.empty()) parsePending();

    if (inTitlePage && parseTitlePage()) return;

    if (parseSection() || parseForcedAction() || parseForcedSceneHeading() || parseForcedCharacter() ||
        parseForcedTransition() || parsePageBreak() || parseLyrics() || parseSynopsis() ||
        parseCenteredText() || parseSceneHeading() || parseTransition() || parseParenthesis() ||
        parseCharacter() || parseDialogue()) {
        return;
    }

    parseAction();
}

void Parser::finalizeParsing() {
    line = "";
    lineTrim = "";
    parsePending();
}

void Parser::addElement(std::shared_ptr<Element> element) {

    auto lastElement = getLastElement();

    if (element->getType() == ElementType::ACTION && element->isEmpty() &&
        !std::dynamic_pointer_cast<Action>(element)->isCentered()) {

        inDialogue = false;

        if (lastElement && lastElement->getType() == ElementType::ACTION) {
            padActions.push_back(std::dynamic_pointer_cast<Action>(element));
            return;
        }
        return;
    }

    if (element->getType() == ElementType::ACTION && !padActions.empty()) {

        if (mergeActions && lastElement && lastElement->getType() == ElementType::ACTION && 
            !std::dynamic_pointer_cast<Action>(lastElement)->isCentered()) {
 
            for (const auto& padAction : padActions) {
                lastElement->appendLine(padAction->getTextRaw());
            }

        } else {
            for (const auto& padAction : padActions) {
                script->addElement(padAction);
            }

        }
    }
    
    padActions.clear();

    if (mergeActions && element->getType() == ElementType::ACTION &&
        !std::dynamic_pointer_cast<Action>(element)->isCentered()) {

        if (lastElement && lastElement->getType() == ElementType::ACTION &&
            !std::dynamic_pointer_cast<Action>(lastElement)->isCentered()) {

            lastElement->appendLine(element->getTextRaw());
            return;
        }

    }

    script->addElement(element);

    inDialogue = element->getType() == ElementType::CHARACTER || element->getType() == ElementType::PARENTHESIS ||
                 element->getType() == ElementType::DIALOGUE;
}

std::shared_ptr<Element> Parser::getLastElement() {
    if (script->getElements().empty()) return nullptr;
    return script->getElements().back();
}

void Parser::parsePending() {
    for (const auto& pendingItem : pending) {
        if (pendingItem->type == ElementType::TRANSITION) {
            if (lineTrim.empty() || lineTrim.find_first_not_of(" \t") == std::string::npos) {
                addElement(pendingItem->element);
            } else {
                addElement(pendingItem->backup);
            }
        } else if (pendingItem->type == ElementType::CHARACTER) {
            if (!lineTrim.empty() && lineTrim.find_first_not_of(" \t") != std::string::npos) {
                addElement(pendingItem->element);
            } else {
                addElement(pendingItem->backup);
            }
        }
    }
    pending.clear();
}

bool Parser::parseBoneyard() {

    size_t open = line.find("/*");
    size_t close = line.find("*/", (open != std::string::npos) ? open : 0);
    size_t lastTag = std::string::npos;

    // Handle in-line boneyards
    while (open != std::string::npos && close != std::string::npos && close > open) {
        // Extract boneyard content
        std::string boneyardText = line.substr(open + 2, close - open - 2);
        script->addBoneyard(std::make_shared<Boneyard>(boneyardText));

        // Replace boneyard content with a tag
        std::string tag = "/*" + std::to_string(script->getBoneyards().size() - 1) + "*/";
        line = line.substr(0, open) + tag + line.substr(close + 2);

        // Update position of lastTag
        lastTag = open + tag.length();

        // Search for the next pair of delimiters
        open = line.find("/*", lastTag);
        close = line.find("*/", lastTag);
    }

    // Check for entering boneyard content
    if (!boneyard) {
        size_t idx = line.find("/*", (lastTag != std::string::npos) ? lastTag : 0);
        if (idx != std::string::npos) {
            lineBeforeBoneyard = line.substr(0, idx);
            boneyard = std::make_shared<Boneyard>(line.substr(idx + 2));
            return true;
        }
    } else {
        // Check for end of boneyard content
        size_t idx = line.find("*/", (lastTag != std::string::npos) ? lastTag : 0);
        if (idx != std::string::npos) {

            // Append content and close the boneyard
            boneyard->appendLine(line.substr(0, idx));
            script->addBoneyard(boneyard);

            // Replace with a tag
            std::string tag = "/*" + std::to_string(script->getBoneyards().size() - 1) + "*/";
            line = lineBeforeBoneyard + tag + line.substr(idx + 2);

            // Reset state
            lineBeforeBoneyard.clear();
            boneyard = nullptr;
        } else {
            // Still in boneyard
            boneyard->appendLine(line);
            return true;
        }
    }

    return false;
}

bool Parser::parseNotes() {

    size_t open = line.find("[[");
    size_t close = line.find("]]", (open != std::string::npos) ? open : 0);
    size_t lastTag = std::string::npos;

    while (open != std::string::npos && close != std::string::npos && close>open) {
        // Extract note text
        std::string noteText = line.substr(open + 2, close - open - 2);

        // Add the note to the script
        script->addNote(std::make_shared<Note>(noteText));

        // Replace note with a tag
        std::string tag = "[[" + std::to_string(script->getNotes().size() - 1) + "]]";
        line = line.substr(0, open) + tag + line.substr(close + 2);

        // Update lastTag position
        lastTag = open + tag.length();

        // Find the next set of delimiters
        open = line.find("[[", lastTag);
        close = line.find("]]", lastTag);
    }

    if (!note) {
        // Start a new note
        size_t idx = line.find("[[", (lastTag != std::string::npos) ? lastTag : 0);
        if (idx != std::string::npos) {
            lineBeforeNote = line.substr(0, idx);
            note = std::make_shared<Note>(line.substr(idx + 2));
            line = lineBeforeNote;
            return true;
        }
    } else {
        // End or continue an existing note
        size_t idx = line.find("]]", (lastTag != std::string::npos) ? lastTag : 0);
        if (idx != std::string::npos) {
            // End of note found
            note->appendLine(line.substr(0, idx));
            script->addNote(note);

            std::string tag = "[[" + std::to_string(script->getNotes().size() - 1) + "]]";
            line = lineBeforeNote + tag + line.substr(idx + 2);
            lineBeforeNote = "";
            note = nullptr;
        } else if (line=="") {
            // End of note due to line break
            script->addNote(note);

            std::string tag = "[[" + std::to_string(script->getNotes().size() - 1) + "]]";
            line = lineBeforeNote + tag;
            lineBeforeNote = "";
            note = nullptr;
        } else {
            // Still in note content
            note->appendLine(line);
            return true;
        }
    }

    return false;
}

bool Parser::parseTitlePage() {
    static const std::regex regexTitleEntry(R"(^\s*([A-Za-z0-9 ]+?)\s*:\s*(.*?)\s*$)");
    static const std::regex regexTitleMultilineEntry(R"(^( {3,}|\t))");

    std::smatch match;
    if (std::regex_match(line, match, regexTitleEntry)) {
        std::string key = match[1].str();
        std::string value = match[2].str();

        script->addTitleEntry(std::make_shared<TitleEntry>(key, value));
        multiLineTitleEntry = value.empty();
        return true;
    }

    if (multiLineTitleEntry && std::regex_search(line, match, regexTitleMultilineEntry)) {
        if (!script->getTitleEntries().empty()) {
            script->getTitleEntries().back()->appendLine(line);
        }
        return true;
    }

    inTitlePage = false;
    return false;
}

bool Parser::parseSection() {
    if (lineTrim.rfind("###", 0) == 0) {
        addElement(std::make_shared<Section>(trim(lineTrim.substr(3)), 3));
        return true;
    }

    if (lineTrim.rfind("##", 0) == 0) {
        addElement(std::make_shared<Section>(trim(lineTrim.substr(2)), 2));
        return true;
    }

    if (lineTrim.rfind("#", 0) == 0) {
        addElement(std::make_shared<Section>(trim(lineTrim.substr(1)), 1));
        return true;
    }

    return false;
}

bool Parser::parseForcedAction() {
    if (lineTrim.rfind("!", 0) == 0) {
        addElement(std::make_shared<Action>(lineTrim.substr(1), true));
        return true;
    }
    return false;
}

bool Parser::parseForcedSceneHeading() {

        // Regex to match scene headings
    static const std::regex regexHeading(R"(^\.[a-zA-Z0-9])");

    // Check if the trimmed line matches the regex
    std::smatch match; // Holds the match result
    if (std::regex_search(lineTrim, match, regexHeading)) {
        auto heading = decodeHeading(lineTrim.substr(1));
        if (heading) {
            addElement(std::make_shared<SceneHeading>(heading->first, heading->second, true));
            return true;
        }
    }
    return false;
}

bool Parser::parsePageBreak() {
    if (lineTrim.find("===") != std::string::npos) {
        addElement(std::make_shared<FountainPageBreak>());
        return true;
    }
    return false;
}

bool Parser::parseForcedTransition() {
    if (lineTrim.starts_with(">") && !lineTrim.ends_with("<")) {
        // Add a forced FountainTransition element with trimmed content
        addElement(std::make_shared<Transition>(trim(lineTrim.substr(1)), true));
        return true;
    }

    return false;
}

bool Parser::parseLyrics() {
    if (lineTrim.starts_with("~")) {
        // Create and add a FountainLyric element
        addElement(std::make_shared<Lyric>(trim(lineTrim.substr(1))));
        return true;
    }
    return false;
}

bool Parser::parseSynopsis() {
    static const std::regex synopsisRegex(R"(^=(?!\=))"); // Matches a single '=' not followed by another '='
    
     std::smatch match;
    if (std::regex_search(lineTrim, match, synopsisRegex)) {
        // Create and add a FountainSynopsis element
        addElement(std::make_shared<Synopsis>(trim(lineTrim.substr(1))));
        return true;
    }
    return false;
}

bool Parser::parseDialogue() {
    auto lastElement = getLastElement();

    // If last element is CHARACTER or PARENTHESIS and line is not empty
    if (lastElement != nullptr && !line.empty() &&
        (lastElement->getType() == ElementType::CHARACTER || lastElement->getType() == ElementType::PARENTHESIS)) {
        addElement(std::make_shared<Dialogue>(lineTrim));
        return true;
    }

    // If last element is DIALOGUE
    if (lastElement != nullptr && lastElement->getType() == ElementType::DIALOGUE) {
        // Handle continuation after an empty line
        if (lastLineEmpty && !lastLine.empty()) {
            if (mergeDialogue) {
                lastElement->appendLine("");
                lastElement->appendLine(lineTrim);
            } else {
                addElement(std::make_shared<Dialogue>(""));
                addElement(std::make_shared<Dialogue>(lineTrim));
            }
            return true;
        }

        // Handle continuation on the same line
        if (!lastLineEmpty && !lineTrim.empty()) {
            if (mergeDialogue) {
                lastElement->appendLine(lineTrim);
            } else {
                addElement(std::make_shared<Dialogue>(lineTrim));
            }
            return true;
        }
    }

    return false;
}

std::optional<Parser::CharacterInfo> Parser::decodeCharacter(const std::string& line) {
    // Regex to match "(CONT'D)"
    std::string noContLine = replaceAll(line, "(CONT'D)", "");
    noContLine = replaceAll(noContLine, "(CONT’D)", "");
    noContLine = trim(noContLine);

    // Regex to match a character line
    static const std::regex regexCharacter(R"(^([^(\^]+?)\s*(?:\((.*)\))?(?:\s*\^\s*)?$)");
    std::smatch match;

    if (std::regex_match(noContLine, match, regexCharacter)) {
        std::string name = match[1].str();
        std::optional<std::string> extension = match[2].matched ? std::optional(match[2].str()) : std::nullopt;
        bool dual = noContLine.back() == '^';

        // Return a populated CharacterInfo struct
        return CharacterInfo{name, extension, dual};
    }
    return std::nullopt; // No match found
}

bool Parser::parseCharacter() {
    std::string noContLineTrim = replaceAll(lineTrim, "(CONT'D)", "");
    noContLineTrim = replaceAll(noContLineTrim, "(CONT’D)", "");
    noContLineTrim = trim(noContLineTrim);

    // Regex to identify character lines
    static const std::regex regexCharacter(R"(^([A-Z][^a-z]*?)\s*(?:\(.*\))?(?:\s*\^\s*)?$)");
    if (lastLineEmpty && std::regex_match(noContLineTrim, regexCharacter)) {
        auto characterOpt = decodeCharacter(noContLineTrim); // Decode the character line
        if (characterOpt) {
            auto character = *characterOpt;

            // Add a new PendingElement to the pending vector
            pending.push_back(std::make_shared<PendingElement>(PendingElement{
                ElementType::CHARACTER,
                std::make_shared<Character>(noContLineTrim, character.name, character.extension, character.dual),
                std::make_shared<Action>(lineTrim)
            }));

            return true;
        }
    }

    return false;
}

bool Parser::parseTransition() {
    // Regex to match transition lines (e.g., "FADE TO:" or similar)
    static const std::regex regexTransition(R"(^\s*(?:[A-Z\s]+TO:)\s*$)");

    // Check if the line matches the regex and if the last line was empty
    if (std::regex_match(lineTrim, regexTransition) && lastLineEmpty) {
        // Add a new PendingElement to the pending vector
        pending.push_back(std::make_shared<PendingElement>(PendingElement{
            ElementType::TRANSITION,
            std::make_shared<Transition>(lineTrim),
            std::make_shared<Action>(lineTrim)
        }));
        return true;
    }

    return false;
}

bool Parser::parseParenthesis() {
    // Regex to match parenthesis lines
    static const std::regex regexParenthesis(R"(^\s*\((.*)\)\s*$)");

    // Match the current line against the regex
    std::smatch match;
    if (std::regex_match(line, match, regexParenthesis)) {
        auto lastElement = getLastElement();

        // Check if the match was successful, we're in dialogue, and the last element is valid
        if (inDialogue && lastElement != nullptr &&
            (lastElement->getType() == ElementType::CHARACTER || lastElement->getType() == ElementType::DIALOGUE)) {
            // Add a new FountainParenthesis element
            addElement(std::make_shared<Parenthesis>(match[1].str()));
            return true;
        }
    }

    return false;
}

bool Parser::parseCenteredText() {
    // Check if lineTrim starts with ">" and ends with "<"
    if (lineTrim.starts_with(">") && lineTrim.ends_with("<")) {
        // Extract the content between ">" and "<"
        std::string content = lineTrim.substr(1, lineTrim.length() - 2);

        // Create a new FountainAction element and mark it as centered
        auto centeredElement = std::make_shared<Action>(content);
        centeredElement->setCentered(true);

        // Add the centered element to the script
        addElement(centeredElement);
        return true;
    }

    return false;
}

bool Parser::parseSceneHeading() {
    // Regex to match scene headings
    static const std::regex regexHeading(
        R"(^\s*((INT|EXT|EST|INT\.\/EXT|INT\/EXT|I\/E)(\.|\s))|(FADE IN:\s*))",
        std::regex::icase
    );

    // Check if the trimmed line matches the regex
     std::smatch match; // Holds the match result
    if (std::regex_search(lineTrim, match, regexHeading)) {
        auto headingOpt = decodeHeading(lineTrim); // Decode the heading text and optional scene number
        if (headingOpt) {
            const auto& [text, sceneNum] = *headingOpt; // Extract text and scene number
            addElement(std::make_shared<SceneHeading>(text, sceneNum));
        }
        return true;
    }

    return false;
}

bool Parser::parseForcedCharacter() {
    // Check if the line starts with "@"
    if (lineTrim.starts_with("@")) {
        // Remove the "@" prefix and trim the remaining string
        std::string trimmedLine = lineTrim.substr(1);
        trimmedLine = trim(trimmedLine);

        // Decode the character details
        auto characterOpt = decodeCharacter(trimmedLine);
        if (characterOpt) {
            auto character = *characterOpt;

            // Create and add a FountainCharacter element
            addElement(std::make_shared<Character>(trimmedLine, character.name, character.extension, character.dual));
            return true;
        }
    }

    return false;
}

void Parser::parseAction() {
    addElement(std::make_shared<Action>(line));
}

std::optional<std::pair<std::string, std::optional<std::string>>> Parser::decodeHeading(const std::string& line) {
    std::regex regex(R"((.*?)(?:\s*#([a-zA-Z0-9\-.]+?)#)?)");
    std::smatch match;

    if (std::regex_match(line, match, regex)) {
        std::string text = match[1];
        std::optional<std::string> sceneNum = match[2].matched ? std::optional(match[2].str()) : std::nullopt;
        return std::make_pair(text, sceneNum);
    }
    return std::nullopt;
}

} // namespace Fountain